chore(docs): add react integration documentation (#321)
elastic/apm-agent-rum-js: docs/react-integration.asciidoc

[[react-integration]]
=== React integration

> This document covers how to use React integration with Real User Monitoring JavaScript agent.

Reviewer: ```suggestion
This section covers how to use Real User Monitoring JavaScript agent on web applications using React library.
```

Please see our <<getting-started, Getting started guide>> for configuring the Real User Monitoring agent.

[[installing-react-integration]]
==== Installing Elastic APM React package

Install the `@elastic/apm-rum-react` package as a dependency to your application:

[source,bash]
----
npm install @elastic/apm-rum-react --save
----

> ==== Instrumenting your React application

Reviewer 2: Let's add float tags to these headers. They'll render correctly now, but if we add anchors later on they will break 

```suggestion
[float]
==== Instrumenting your React application
```

The React integration package provides two approaches to instrumenting your application:

> ===== Instrumenting application routes

Reviewer 2: ```suggestion
[float]
===== Instrumenting application routes
```

To instrument the application routes, you can use `ApmRoute` component provided in the package. 
`ApmRoute` creates a transaction that has the path of the `Route` as its name and has `route-change`
as its type.

> NOTE: Currently `ApmRoute` only supports applications using `react-router` library.

Reviewer: link to the react-router lib would be good addition.

First you should import `ApmRoute` from the `@elastic/apm-rum-react` package:

[source,js]
----
import { ApmRoute } from '@elastic/apm-rum-react'
----

Then, you should replace `Route` components from the `react-router` library
with `ApmRoute`. You can use `ApmRoute` in any of the routes that you would like to monitor,
 therefore, you don't have to change all of your routes:

[source,js]
----

> class App extends React.Component {

Reviewer: nit: change indent size to 2 instead of 4

    render() {
        return (
            <div>
                <ApmRoute
                    exact
                    path="/"
                    component={() => (
                        <Redirect
                            to={{
                                pathname: '/home'
                            }}
                        />
                    )}
                />
                <ApmRoute path="/home" component={HomeComponent} />
                <Route path="/about" component={AboutComponent} />
            </div>
        )
    }
}
----

> ===== Instrumenting individual React components

Reviewer 2: ```suggestion
[float]
===== Instrumenting individual React components
```

First you should import `withTransaction` from the `@elastic/apm-rum-react` package:

[source,js]
----
import { withTransaction } from '@elastic/apm-rum-react'
----

Then, you can use `withTransaction` to decorate your React components,
either using the decorator notation:

[source,js]
----

> @withTransaction('AboutComponent', 'component')

Reviewer: Let's remove the decorator pattern here, Decorators is still at stage 2 and we should not be encouraging people to use that. If we do that, we have to also tell users how to enable that as well. Lets keep it simple and only add function example.

Author: I don't think of it as encouraging people to use it to be honest, it's simply a choice users can make. I've seen plenty of setups that are already using decorators, similar to how most of the syntax in a JSX file is either not standard at all or needs transpiling. I will add a note about on usage thought.

Author: We decide to remove the decorator usage for now, we can revisit once we support typescript.

class AboutComponent extends React.Component { }
----

Or, using `withTransaction` as a function:

[source,js]
----
class AboutComponent extends React.Component { }
export default withTransaction('AboutComponent', 'component')(AboutComponent)
----

NOTE: `withTransaction` accepts two parameters, "transaction name" and "transaction type". 

> If these parameters are not provided the defaults documented in <<transaction-api, Transaction API>> will be used.

Reviewer 2: ```suggestion
If these parameters are not provided, the defaults documented in <<transaction-api, Transaction API>> will be used.
```
